core/namespace: fix issue in PrivateTmp=disconnected vs var.mount

yuwata opened the pull request:
Fixes #37258.
Replaces #37259.

YHNdnzj commented on src/core/namespace.c:
```diff
@@ -1790,6 +1792,16 @@ static int apply_one_mount(
         }
 
+        case MOUNT_EMPTY_DIR_WHEN_EMPTY:
+                r = dir_is_empty(mount_entry_path(m), /* ignore_hidden_or_backup= */ false);
```
I'm not convinced. This basically means what the service gets for /var/ is racy for no good. Any future mounts cannot be propagated into namespace anymore. And upon later restart they might get completely different stuff...

Like, if `RootImage=` is used, what's mounted in /var/ _is_ stable, regardless of PrivateTmp=. Now, there's no guarantee that services on host, even if not explicitly ordered after /var/, doesn't need to access it. It's fair to assume that /var/ will be mounted later on, and they can monitor until that happens and operate on that. In fact, networkd-persistent-storage is backed by a similar idea, just we used fd passing so this is a non-issue.

Frankly, let's not be overzealous and make up problems.

YHNdnzj replied:
Even then we're still left with the race part. I don't see the rationale behind this...

YHNdnzj replied:
This check is not really useful btw. If /var/ is not going to be a mount yet still is empty, we'd mount a tmpfs there pointlessly.

YHNdnzj commented on src/core/namespace.c:
```diff
@@ -1907,6 +1919,11 @@ static int apply_one_mount(
         return 1;
 }
 
+static bool propagate_attribute_to_submounts(const MountEntry *m) {
```
call it should_propagate_?

YHNdnzj commented on units/systemd-journal-remote.service.in:
```diff
@@ -20,7 +20,7 @@ MemoryDenyWriteExecute=yes
 PrivateDevices=yes
 PrivateNetwork=yes
-PrivateTmp=yes
+PrivateTmp=disconnected
```
this is wrong. The premise behind disconnected is that any quota established on `/tmp/` is taken away, along with the ability to inspect `/tmp/` of a service. We hence need to be very careful with enabling this, i.e. only when the dependency on `systemd-tmpfiles-setup.service` is strictly undesirable, but nothing else.

Please drop it.

YHNdnzj replied:
I mean, that's the whole deal for making `PrivateTmp=yes` backed by `/tmp/`, so that quota or setups where `/tmp/` is NOT a tmpfs is respected.

yuwata replied:
Hmm, quota... OK. Dropped.

yuwata commented:
BTW, is it necessary to enable PrivateTmp= in resolved or friends, especially when ProtectSystem=strict is set?

poettering commented:
so this is obviously a mess. 

Here are two ideas:

1. Maybe there should be two distinct enum values, maybe "PrivateTmp=disconnected" and "PrivateTmp=disconnected-with-var" or so. One only deals with /tmp, the other also deals with /var/tmp, and adds the necessary dep. Then we can communicate clearly in the docs what actually happens.

2. Maybe in PrivateTmp=disconnected mode we should not bother with /var/tmp at all, and instead set the TMPDIR env var explicitly to /tmp/. my understanding of $TMPDIR is after all that it redirects *both* uses to the configured dir. i.e. a correctly written program that uses /tmp/ or /var/tmp should honour $TMPDIR in all cases if its set.

And maybe even do a combination of this...

What do you think?

poettering commented:
another idea: make (and document) that PrivateTmp=disconnected has two distinct behaviours depending if DefaultDependencies= is true or false. If it is true then we'll also do /var/tmp handling, and imply a dep on /var (that dep is redundant of course). If it is false we'll not do /var/tmp handling, and not add any dep (and possibly set $TMPDIR).

poettering commented:
I agree with @YHNdnzj btw that this race-prone behaviour what to do with /var/ (where sometimes you end up with one thing, sometimes with another) is not ideal. I'd very much prefer if the behaviour depends on explicit config only, i.e. unit file settings

yuwata commented:
> another idea: make (and document) that PrivateTmp=disconnected has two distinct behaviours depending if DefaultDependencies= is true or false. If it is true then we'll also do /var/tmp handling, and imply a dep on /var (that dep is redundant of course). If it is false we'll not do /var/tmp handling, and not add any dep (and possibly set $TMPDIR).

I like this one.

bluca commented:
> another idea: make (and document) that PrivateTmp=disconnected has two distinct behaviours depending if DefaultDependencies= is true or false. If it is true then we'll also do /var/tmp handling, and imply a dep on /var (that dep is redundant of course). If it is false we'll not do /var/tmp handling, and not add any dep (and possibly set $TMPDIR).

As long as the RootImage/RootDirectory cases are still considered as exceptions, this sounds fine to me

bluca commented:
ie, for my use cases it's really important that when something doesn't run on the rootfs, there are no dependencies added, but it still gets tmp and var/tmp handled

YHNdnzj commented:
@yuwata consider cherry-picking the first commit in my PR here?

poettering commented:
yeah, so I guess the overall rule should then be:

1. If DefaultDependencies= is turned off and RootImage=/RootDirectory= are not set → do not bother with /var/tmp/ at all, and set $TMPDIR=/tmp env var.

2. Otherwise mount a tmpfs to /var/tmp, and do not set $TMPDIR.

Everyone on board with that?

yuwata commented:
Done, except for updating the man page. If you'd like, please review or test.

Cornelicorn commented:
> Done, except for updating the man page. If you'd like, please review or test.

Tested it, solves #37258

YHNdnzj commented on src/core/unit.c:
```diff
@@ -1281,6 +1281,12 @@ int unit_add_exec_dependencies(Unit *u, ExecContext *c) {
                         return r;
+        } else if (c->private_var_tmp == PRIVATE_TMP_DISCONNECTED && !exec_context_with_rootfs(c)) {
+                /* Even if PrivateTmp=disconnected, we still require /var/tmp/ mountpoint to be present,
+                 * i.e. /var/ needs to be mounted. */
```
Mention `unit_patch_contexts`?

YHNdnzj commented on src/core/unit.c:
```diff
@@ -1281,6 +1281,12 @@ int unit_add_exec_dependencies(Unit *u, ExecContext *c) {
                 r = unit_add_dependency_by_name(u, UNIT_AFTER, SPECIAL_TMPFILES_SETUP_SERVICE, true, UNIT_DEPENDENCY_FILE);
```
assert on private_var_tmp also being CONNECTED here?

yuwata commented:
Thank you for the review and test. The man page is also updated. PTAL.

Cornelicorn commented:
Btw, it would be very nice to get a stable backport for this,so this can land in Debian 13 (v257).

poettering commented on src/core/execute.h:
```diff
@@ -330,6 +330,7 @@ struct ExecContext {
         int bind_log_sockets;
         int memory_ksm;
         PrivateTmp private_tmp;
+        PrivateTmp private_var_tmp;
```
hmm, why is this a variable of its own?

why not instead have a helper that determines this on the fly?

i.e. exec_context_private_var_tmp() that checks private_tmp, and the rootdir/rootimage stuff and returns the right value?

poettering replied:
i think it might be ok to keep this two separate fields in NamespaceParameters, but in ExecContext make this one field + one accessor

yuwata replied:
I initially tried that, but it depends on Unit.default_dependencies, and mapping from ExecContext to Unit is not straight forward. Moreover, the contents of Unit is not serialized to systemd-executor. So, I decided to use a variable.

poettering replied:
hmm, i see the problem, but can we at least initialize private_var_tmp to _PRIVATE_TMP_INVALID then, and then in unit_patch_contexts() set it to the final value, based on private_tmp, in all cases?

i.e. i'd prefer if we can always discern the cases private_var_tmp is off, from private_var_tmp is just not patched in yet, for whatever reason, so that any issues with not initializing this for some reason are caught.

also please add a comment explaining that private_var_tmp is not really an independent option, but always initialized from private_tmp

yuwata commented on src/core/unit.c:
```diff
@@ -4325,6 +4334,15 @@ int unit_patch_contexts(Unit *u) {
+                        ec->private_var_tmp = PRIVATE_TMP_NO;
+                else
+                        /* Otherwise, use the same setting for /tmp/. */
+                        ec->private_var_tmp = ec->private_tmp;
```
See the reply above.

poettering commented:
> What about turning this around? If `RootDirectory=`/`RootImage=` are unset it implies `RequiresMountsFor=/var` (unless `DefaultDependencies=no`), and if `RequiresMountsFor=/var` is not present the behavior with `$TMPDIR...` is used.

Note that DefaultDependencies=yes already implies RequiresMountsFor=/var, though indirectly via local-fs.target and basic.target. 

> Only difference: Even with `DefaultDependencies=no` once could still explicitly use the other behavior by manually adding back `RequiresMountsFor=/var`. This feels more natural

I'd be fine to tweak the logic so that an explicit RequiresMountsFor=/var is also a reason to do the full /var/tmp/ thing.

yuwata commented:
Addressed suggestions by @poettering and @septatrix.
The change since last review is not small, unsetting the green label.
Please take another look. Thank you!

YHNdnzj commented on src/core/unit.c:
```diff
@@ -4256,6 +4262,53 @@ static int unit_verify_contexts(const Unit *u) {
+        return PRIVATE_TMP_NO;
+}
+
+
```
spurious newline

YHNdnzj commented on src/core/unit.c:
```diff
@@ -4256,6 +4262,53 @@ static int unit_verify_contexts(const Unit *u) {
         return 0;
 }
 
+static PrivateTmp unit_get_private_var_tmp(Unit *u, ExecContext *c) {
```
const'ify u and c?

YHNdnzj commented on src/core/unit.c:
```diff
@@ -4256,6 +4262,53 @@ static int unit_verify_contexts(const Unit *u) {
+                return PRIVATE_TMP_DISCONNECTED;
+
+        /* Even if DefaultDependencies=no, enable disconnected tmpfs when
+         * RequiresMountsFor=WantsMountsFor=/var/ is explicitly set. */
```
```suggestion
         * RequiresMountsFor=/WantsMountsFor=/var/ is explicitly set. */
```

YHNdnzj commented on src/core/unit.c:
```diff
@@ -4256,6 +4262,53 @@ static int unit_verify_contexts(const Unit *u) {
+
+        if (!unit_has_dependency(u, UNIT_ATOM_AFTER, m))
+                return PRIVATE_TMP_NO;
+
+        if (unit_has_dependency(u, unit_dependency_to_atom(UNIT_REQUIRES), m))
+                return PRIVATE_TMP_DISCONNECTED;
+
+        if (unit_has_dependency(u, unit_dependency_to_atom(UNIT_WANTS), m))
+                return PRIVATE_TMP_DISCONNECTED;
```
hmm, i'd probably not bother with catching Wants=/Requires= here. Any setup with `var.mount` around and carries `After=var.mount` for the unit is 99.99% going to have `/var/` mounted by local-fs.target. The rest 0.01% is outright broken, because `/var/` will be pulled in by any unit with StateDirectory= and friends or PrivateTmp=yes. We don't need to care about them if you ask me.

YHNdnzj replied:
and there's BindsTo=. So the current list is non-exhaustive

yuwata replied:
But, RequiresMountsFor= is exactly the same as After= and Requires= especially for /var/. So, I do not think we should take one of them. If we take RequiresMountsFor=, we should also take After= and Requires=.

There is no BindsMountsFor=. So, I think it is OK to not consider BindsTo=, even though it should work. People can specify both BindsTo=var.mount and RequiresMountsFor=/var/ at the same time.

YHNdnzj replied:
> But, RequiresMountsFor= is exactly the same as After= and Requires= especially for /var/. So, I do not think we should take one of them. If we take RequiresMountsFor=, we should also take After= and Requires=.

They aren't though. `Requires=` mandates var.mount to exist, while mounts for deps first checks if it's a mountpoint at all. IOW, `RequiresMountsFor=` is inherently weaker than Requires=, and it tries to ensure /var/ is around, not about var.mount being _mounted_.

(a quick search reveals several occurrences of standalone `After=var.mount` btw: https://grep.app/search?case=true&q=After%3Dvar.mount , exactly because of the reasons stated above)

YHNdnzj replied:
I mean, they probably come from the ancient era before RequiresMountsFor=. But the fact that `After=var.mount` exists kinda suggests their expectation on `/var/`. So we can just enable `/var/` for them...

yuwata replied:
Even though, the net effect for the current problem should be the same. If a unit has Requires= and After= var.mount, then we can safely mount tmpfs on /var/tmp/.

YHNdnzj replied:
With After=var.mount it should be safe too... Or, please check for UNIT_ATOM_PULL_IN_START and UNIT_ATOM_PULL_IN_START_IGNORED. I really think Requires= and Wants= is not special here.

yuwata replied:
Done. PTAL.

YHNdnzj commented on src/core/execute-serialize.c:
```diff
@@ -1735,6 +1735,11 @@ static int exec_context_serialize(const ExecContext *c, FILE *f) {
         if (r < 0)
                 return r;
 
+        assert(c->private_var_tmp >= 0 && c->private_var_tmp < _PRIVATE_TMP_MAX);
```
no need for such specific knowledge in serialization code? it's validated multiple times elsewhere after all.

yuwata replied:
This ensures private_var_tmp is explicitly set before invoking the executor.
Of course, the same check is done at several more places, so redundant. But this is just an assertion and we have so many trivial assertions...

YHNdnzj replied:
> But this is just an assertion and we have so many trivial assertions...

Sure, but currently the serialization code carries none of those. but it's fine either way.

YHNdnzj reviewed (commented):
LGTM overall

YHNdnzj commented on src/core/namespace.c:
```diff
@@ -731,6 +731,107 @@ static int append_tmpfs_mounts(MountList *ml, const TemporaryFileSystem *tmpfs,
+        assert(ml);
+        assert(p);
+
+        /* When DynamicUser=yes enforce that /tmp/ and /var/tmp/ are disconnected from the host's
+         * directories, as they are world writable and ephemeral uid/gid will be used. */
```
not your fault (mine actually), but this should be moved to unit_patch_contexts after PrivateTmp=disconnected is exposed

yuwata replied:
There is already similar comments. I will remove this one.

YHNdnzj commented on src/core/unit.c:
```diff
@@ -4256,6 +4262,55 @@ static int unit_verify_contexts(const Unit *u) {
+        if (!m)
+                return PRIVATE_TMP_NO;
+
+        if (!unit_has_dependency(u, UNIT_ATOM_AFTER, m))
+                return PRIVATE_TMP_NO;
+
+        if (unit_has_dependency(u, UNIT_ATOM_PULL_IN_START, m))
+                return PRIVATE_TMP_DISCONNECTED;
+
+        if (unit_has_dependency(u, UNIT_ATOM_PULL_IN_VERIFY, m))
+                return PRIVATE_TMP_DISCONNECTED;
+
+        if (unit_has_dependency(u, UNIT_ATOM_PULL_IN_START_IGNORED, m))
+                return PRIVATE_TMP_DISCONNECTED;
```
I'd place these in a single if block to make the interaction more clear and reduce a bit of noise.

YHNdnzj commented:
ugh, it appears that some of our unit tests sidestep unit_patch_contexts for some reason...

yuwata commented:
Several non-small changes are added. Could you review again?

YHNdnzj commented on src/core/unit.c:
```diff
@@ -1896,8 +1902,16 @@ int unit_start(Unit *u, ActivationDetails *details) {
+        if (u->load_state != UNIT_LOADED) {
+                if (MANAGER_IS_TEST_RUN(u->manager)) {
+                        assert(u->load_state == UNIT_STUB);
+                        u->load_state = UNIT_LOADED;
```
This doesn't seem right? There are also "fully-charged" test runs that go through usual unit load routines, e.g. analyze?

yuwata replied:
Yeah, I concerned that. Also, e.g. test-execute does load units. But in that case, the state should be already in UNIT_LOADED, so I concluded that this is not a big matter.
If you do not like the change, I will drop the change, and call unit_patch_contexts() everywhere we need.

yuwata replied:
OK. Dropped the change, and unit_patch_contexts() is called in tests where necessary.

YHNdnzj commented on src/core/mount.c:
```diff
@@ -1855,10 +1855,13 @@ static int mount_setup_existing_unit(
                 u->load_error = 0;
 
+                unit_add_to_load_queue(u);
+
```
Hmm, the JUST_CHANGED branch below can be changed to else if, hence everything will be reloaded when dispatching load queue here?

yuwata replied:
That's a kind of optimization. Done.

YHNdnzj commented:
LGTM, too

keszybz commented:
This was just merged… I'm leaving it out of the batch for v257.6 to let it bake a bit. It's quite a lot of changes.

mnsgs commented:
Any chance of seeing this (and associated https://github.com/systemd/systemd/pull/37637) in a v257.8?
